File: abc264/e2.py
import sys
sys.setrecursionlimit(2*10**5 + 1)

class UnionFind:
    def __init__(self,N):
        self.__parents_or_size = [-1 for i in range(N)]
    
    def root(self,x):
        if self.__parents_or_size[x] < 0:
            return x
        else:
            self.__parents_or_size[x] = self.root(self.__parents_or_size[x])
            return self.__parents_or_size[x]
    
    def size(self,x):
        return -self.__parents_or_size[self.root(x)]
    
    def same(self,x,y):
        return self.root(x) == self.root(y)
    
    def unite(self,x,y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return
        self.__parents_or_size[x] += self.__parents_or_size[y]
        self.__parents_or_size[y] = x



N,M,E = map(int,input().split())
uv = []
for i in range(E):
    u,v = map(int,input().split())
    u-=1
    v-=1
    uv.append([u,v])

Q = int(input())
X = [int(input())-1 for i in range(Q)]
X_set = set(X)
electrics = [False for i in range(N)] + [True for i in range(M)]
uf = UnionFind(N+M+1)

cur = 0
for i in range(E):
    
    if i in X_set:
        continue
    u,v = uv[i]
    
    ru,rv = uf.root(u),uf.root(v)
    
    if electrics[ru] and (not electrics[rv]):
        cur += uf.size(v)
        electrics[rv] = True
    
    elif electrics[rv] and (not electrics[ru]):
        cur += uf.size(u)
        electrics[ru] = True

    uf.unite(u, v)
ans = [cur]
for i in range(Q-1,0,-1):
    u,v = uv[X[i]]
    ru,rv = uf.root(u),uf.root(v)
    
    if electrics[ru] and (not electrics[rv]):
        cur += uf.size(v)
        electrics[rv] = True
    
    elif electrics[rv] and (not electrics[ru]):
        cur += uf.size(u)
        electrics[ru] = True
    ans.append(cur)
    uf.unite(u, v)

ans = reversed(ans)
for a in ans:
    print(a)

File: abc264/e2.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

class UnionFind
{
private:
    vector<long long> parent_or_size;
    long long N;

public:
    UnionFind(long long s)
    {
        N = s;
        parent_or_size.assign(N, -1);
    }
    long long root(long long x)
    {
        if (parent_or_size[x] < 0)
        {
            return x;
        }
        parent_or_size[x] = root(parent_or_size[x]);
        return parent_or_size[x];
    }
    long long is_same(long long x, long long y)
    {
        return root(x) == root(y);
    }
    long long size(long long x)
    {
        return -parent_or_size[root(x)];
    }
    void unite(long long x, long long y)
    {
        long long rx = root(x);
        long long ry = root(y);
        if (rx == ry)
        {
            return;
        }
        parent_or_size[rx] += parent_or_size[ry];
        parent_or_size[ry] = rx;
    }
};

int main()
{
    ll N, M, E;
    cin >> N >> M >> E;
    vector<pair<ll, ll>> uv(E);
    for (ll i = 0; i < E; i++)
    {
        ll u, v;
        cin >> u >> v;
        u--;
        v--;
        uv[i] = {u, v};
    }
    ll Q;
    cin >> Q;
    vector<ll> X(Q);
    set<ll> X_set;
    for (ll i = 0; i < Q; i++)
    {
        cin >> X[i];
        X[i]--;
        X_set.insert(X[i]);
    }

    auto uf = UnionFind(N + M + 1);
    vector<bool> electrocuted(N + M);
    for (ll i = N; i < N + M; i++)
    {
        electrocuted[i] = true;
    }
    ll electrocuted_cities = 0;
    for (ll i = 0; i < E; i++)
    {
        if (X_set.find(i) != X_set.end())
        {
            continue;
        }
        auto u = uv[i].first;
        auto v = uv[i].second;
        if (uf.is_same(u, v))
        {
            continue;
        }
        auto ru = uf.root(u);
        auto rv = uf.root(v);
        if (electrocuted[ru] && !electrocuted[rv])
        {
            electrocuted_cities += uf.size(v);
            electrocuted[rv] = true;
        }
        if (!electrocuted[ru] && electrocuted[rv])
        {
            electrocuted_cities += uf.size(u);
            electrocuted[ru] = true;
        }
        uf.unite(u, v);
    }
    vector<ll> ans = {electrocuted_cities};
    for (ll i = Q - 1; i >= 0; i--)
    {
        auto x = X[i];
        auto u = uv[x].first;
        auto v = uv[x].second;
        if (uf.is_same(u, v))
        {
            ans.push_back(electrocuted_cities);
            continue;
        }
        auto ru = uf.root(u);
        auto rv = uf.root(v);
        if (electrocuted[ru] && !electrocuted[rv])
        {
            electrocuted_cities += uf.size(v);
            electrocuted[rv] = true;
        }
        if (!electrocuted[ru] && electrocuted[rv])
        {
            electrocuted_cities += uf.size(u);
            electrocuted[ru] = true;
        }
        ans.push_back(electrocuted_cities);
        uf.unite(u, v);
    }
    for (ll i = Q-1; i > -1; i--)
    {
        cout << ans[i] << endl;
    }
}